rules/template-indent.js

```diff
@@ -0,0 +1,137 @@
+'use strict';
+const stripIndent = require('strip-indent');
+const {replaceTemplateElement} = require('./fix/index.js');
+
+const MESSAGE_ID_IMPROPERLY_INDENTED_TEMPLATE = 'template-indent';
+const messages = {
+	[MESSAGE_ID_IMPROPERLY_INDENTED_TEMPLATE]: 'Templates should be properly indented. Selector: {{ selector }}',
+};
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => {
+	const options = {
+		tags: ['outdent', 'dedent', 'gql', 'sql', 'html', 'styled'],
+		functions: ['dedent', 'stripIndent'],
+		selectors: [],
+		comments: ['HTML', 'indent'],
+		...context.options[0],
+	};
+
+	options.comments = options.comments.map(comment => comment.toLowerCase());
+
+	const selectors = [
+		...options.tags.map(tag => `TaggedTemplateExpression[tag.name="${tag}"] > .quasi`),
+		...options.functions.map(fn => `CallExpression[callee.name="${fn}"] > .arguments:first-child`),
+		...options.selectors,
+	];
+
+	/** @param {string} selector */
+	const getTemplateLiteralHandler = selector =>
+		/** @param {import('@babel/core').types.TemplateLiteral} node */
+		node => {
+			if (node.type !== 'TemplateLiteral') {
+				return;
+			}
+
+			const delimiter = '__PLACEHOLDER__' + Math.random();
+			const joined = node.quasis.map(q => q.value.raw).join(delimiter);
+
+			if (!joined.includes('\n')) {
+				return;
+			}
+
+			const dedented = stripIndent(joined).trim();
+
+			const startLine = context.getSourceCode().lines[node.loc.start.line - 1];
+			const marginMatch = startLine.match(/^(\s*)\S/);
+			const parentMargin = marginMatch ? marginMatch[1] : '';
+			const tabs = parentMargin.startsWith('\t');
+			const indent = tabs ? '\t' : '  ';
+			const templateMargin = parentMargin + indent;
+
+			const fixed = '\n'
+				+ dedented
+					.split('\n')
+					.map(line => line && line !== '\r' ? templateMargin + line : line)
+					.join('\n')
+					.trimEnd()
+				+ '\n'
+				+ parentMargin;
```

Comment:
sindresorhus have package [`indent-string`](https://github.com/sindresorhus/indent-string), maybe we can use it? And I think it can handle `\r` `\r\n` better. Or maybe just [`redent`](https://github.com/sindresorhus/redent). I havn't test any of them, just an idea. We need tests for `\r` `\r\n` and mixed line breaks.

Reply:
I used indent-string, but something seems to be stripping out `\r` from the input somewhere along the way. I'm hopeful that indent-string and strip-indent do "the right thing", but any ideas of how I can get this working?